### ISSUE TITLE
Supervisor and its child processes freeze on Linux

### ISSUE THREAD

**Issue by TheJamesD** [Supervisor/supervisor#691]
I have supervisor running three different python scripts which daemonize. After they run for some time, I noticed that the scripts freeze, including the Supervisor daemon.

Every time I send a signal to the supervisor process, it will output one more line to the supervisord.log file.

Example if I do the following:

1. Kill a child process of supervisor by doing "kill -9 PID".
2. In the supervisord.log, it will show this: "exited: child1 (terminated by SIGKILL; not expected)"
3. I would now expect Supervisor to try and restart the process because it's an unexpected shutdown, instead nothing happens.
4. I then send a signal to the supervisor daemon process, for example: "kill -s SIGHUP PIDID".
5. I then check the supervisord.log and then supervisor then outputs: "spawned: 'child1' with pid 26001". This is supervisor restarting the child because I sent a SIGKILL. The important thing to note, it was delayed! This should of happened straight after killing the process (step 2).
6. I then send the same SIGHUP signal again, then "supervisord.log" outputs: "received SIGHUP indicating restart request".

It appears that every time I send a signal to the supervisor daemon process, it unbuffers to supervisord.log.

What could cause Supervisor to act like this? Why is it freezing?

I am running Ubuntu 14.04 in a Docker container.

**Comment by mnaberez** [comment]
> I have supervisor running three different python scripts which daemonize. 

Programs run under supervisord must run in the foreground.  See:
http://supervisord.org/subprocess.html#nondaemonizing-of-subprocesses

**Comment by TheJamesD** [comment]
Sorry my mistake, wasn't thinking. The three scripts don't daemonize, they run in the foreground. What could cause the freezing behaviour?

Thanks again for your help.

**Comment by mnaberez** [comment]
What version of Supervisor (`supervisord --version`) and (`python --version`) are you using?

**Comment by TheJamesD** [comment]
Python version: 2.7.6
Supervisor version: 3.0b2

**Comment by TheJamesD** [comment]
Hi Guys,

I am still trying to resolve the Supervisor issue, would it be possible for you to point me in the right direction? Do you have any ideas what could cause behaviour like I explained?

Thanks for your help

**Comment by mnaberez** [comment]
> 1. Kill a child process of supervisor by doing "kill -9 PID".
> 2. In the supervisord.log, it will show this: "exited: child1 (terminated by SIGKILL; not expected)"
> 3. I would now expect Supervisor to try and restart the process because it's an unexpected shutdown, instead nothing happens.

I attempted to reproduce this by doing the following.

Minimal `supervisord.conf`:

    [supervisord]
    logfile = /tmp/supervisord.log
    
    [inet_http_server]
    port = 127.0.0.1:9001
    
    [supervisorctl]
    serverurl = http://127.0.0.1:9001
    
    [rpcinterface:supervisor]
    supervisor.rpcinterface_factory = supervisor.rpcinterface:make_main_rpcinterface
    
    [program:cat]
    command = /bin/cat
    startsecs = 0
    autostart = true
    autorestart = true
  
Start `supervisord` in the foreground using that config:

    $ supervisord --version
    3.1.3
    $ supervisord -n -c /path/to/supervisord.conf

Leave it running, and in a new shell `kill -9` the subprocess:

    $ kill -9 `supervisorctl -c /path/to/supervisord.conf pid cat`

Each time I do this, I see it restarted immediately in the terminal:

    2015-11-18 13:51:55,873 INFO exited: cat (terminated by SIGKILL; not expected)
    2015-11-18 13:51:56,875 INFO spawned: 'cat' with pid 38353
    2015-11-18 13:51:57,877 INFO success: cat entered RUNNING state, process has stayed up for > than 0 seconds (startsecs)

I also see the log update immediately when I watch `tail -f /tmp/supervisord.log`.

I'm unable to duplicate the freezing behavior you described.  The next step would be to try the steps above on your system and see if you get a different result.  If it also works for you, then something else must be wrong, but we need some steps to reproduce it similar to the steps above.

**Comment by TheJamesD** [comment]
Thank you for trying to reproduce the problem. It turns out the issue has nothing to do with Supervisor.

The way I started the Supervisord process is through a Python Subprocess call. The subprocess stdout pipe has a buffer of about 60KB. Once the supervisor output more than 60KB of logs, it caused this weird behaviour. More about it can be found here: http://thraxil.org/users/anders/posts/2008/03/13/Subprocess-Hanging-PIPE-is-your-enemy/

I fixed the issue by not piping the stdout.

Thanks again for your help.
